pkg/gatherers/clusterconfig/machine_configs.go

```diff
@@ -0,0 +1,101 @@
+package clusterconfig
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/insights-operator/pkg/record"
+)
+
+var (
+	sapMachineConfigNameList = map[string]struct{}{
+		"75-worker-sap-data-intelligence": {},
+		"75-master-sap-data-intelligence": {},
+	}
+)
+
+// GatherMachineConfigs collects MachineConfigs definitions. Following data is intentionally removed from the definitions:
+// * `spec.config.storage.files`
+// * `spec.config.passwd.users`
+//
+// The SDI/SAP related MachineConfigs are an exception and are gathered completely
+//
+// Response see https://docs.openshift.com/container-platform/4.7/rest_api/machine_apis/machineconfig-machineconfiguration-openshift-io-v1.html
+//
+// * Location in archive: config/machineconfigs/<name>.json
+// * Id in config: machine_configs
+// * Since versions:
+//   * 4.8+
+func (g *Gatherer) GatherMachineConfigs(ctx context.Context) ([]record.Record, []error) {
+	gatherDynamicClient, err := dynamic.NewForConfig(g.gatherKubeConfig)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return gatherMachineConfigs(ctx, gatherDynamicClient)
+}
+
+func isSAPMachineConfig(mc unstructured.Unstructured) bool {
+	if _, exists := sapMachineConfigNameList[mc.GetName()]; exists {
+		return true
+	}
+
+	for labelName, labelValue := range mc.GetLabels() {
+		if labelName == "workload" && labelValue == "sap-data-intelligence" {
+			return true
+		}
+	}
+
+	for _, ownerRef := range mc.GetOwnerReferences() {
+		if ownerRef.Kind == "ContainerRuntimeConfig" && ownerRef.Name == "sdi-pids-limit" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func gatherMachineConfigs(ctx context.Context, dynamicClient dynamic.Interface) ([]record.Record, []error) {
+	mcList, err := dynamicClient.Resource(machineConfigGroupVersionResource).List(ctx, metav1.ListOptions{})
+	if errors.IsNotFound(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	records := []record.Record{}
+	var errs []error
+	for i := range mcList.Items {
+		mc := mcList.Items[i]
+		// preservce the original behavior - that is gathering of complete SAP machineconfig definitions
+		if !isSAPMachineConfig(mc) {
+			// remove the sensitive content by overwriting the values
+			err := unstructured.SetNestedField(mc.Object, "intentionally overwritten by Insights Operator", "spec", "config", "storage", "files")
```

Comment:
This is the question how to omit the sensitive data. This will replace the `spec.config.storage.files` with the string `intentionally overwritten by Insights Operator`. I can see other options:

- put just empty struct there
- put XXXX to resemble our legacy anonymization

WDYT please?

Reply:
OK so it can't be an empty struct, but I can pass `nil` simply. Any thoughts please @Sergey1011010, @jholecek-rh ?

Reply:
@tremes A preferred solution from the rule development point of view would be:
1. The object remains "valid" after the redaction. As it is a struct , it should remain a struct, not become a string. `nil` is OK.
2. The rule has a way to tell if it was redacted (e.g. gatherers configuration) and that the value cannot be trusted (e.g. a rule checking that it is not empty should not hit on `nil` in that case).

Does it make sense/is it applicable in this context?

Reply:
I am not sure I fully understand your second point (is `nil` still ok or is it in contradiction with point 1 ?:)). What do you mean by `gatherers configuration` please?

Reply:
I think replacing with `nil` would be fine. We also need to mention it in the docs somewhere.

Reply:
OK thanks. It is already redacted here regardless any configuration. The point here is that we know that `spec.config.storage.files` will very likely contain sensitive data (in plain string) as for example certificates and this data would be very difficult to obfuscate (so we can say it's kinda a new approach how to work with sensitive data).

Reply:
Got it. Let's make it `nil` and let's add a note about it in the gatherer's documentation.

Reply:
Thanks. Both are already done. Resolving the conversation.